File: day-1/go/main.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	v1()
	v2()
}

func v1() {
	rawInput, err := os.ReadFile("../input.txt")
	if err != nil {
		return
	}

	var lines = strings.Split(string(rawInput), "\n")

	var distance = 0.0

	var left_array = []int{}
	var right_array = []int{}

	for i := 0; i < len(lines); i++ {
		var number_slices = strings.Split(lines[i], "   ")

		if len(number_slices) < 2 {
			continue
		}

		var left int
		var right int

		left, _ = strconv.Atoi(number_slices[0])
		right, _ = strconv.Atoi(number_slices[1])

		left_array = append(left_array, left)
		right_array = append(right_array, right)

		sort.Ints(left_array)
		sort.Ints(right_array)

	}

	for i := 0; i < len(left_array); i++ {

		var diff = left_array[i] - right_array[i]

		distance += math.Abs(float64(diff))
	}

	fmt.Println("V1:", int32(distance))

}

func v2() {
	rawInput, err := os.ReadFile("../input.txt")
	if err != nil {
		return
	}

	var lines = strings.Split(string(rawInput), "\n")

	var similarity = 0

	var left_map = make(map[int]int)
	var right_map = make(map[int]int)

	for i := 0; i < len(lines); i++ {
		var number_slices = strings.Split(lines[i], "   ")

		if len(number_slices) < 2 {
			continue
		}

		var left int
		var right int

		left, _ = strconv.Atoi(number_slices[0])
		right, _ = strconv.Atoi(number_slices[1])

        left_map[left] = left_map[left] + 1
        right_map[right] = right_map[right] + 1
	}

    fmt.Println(left_map[1])
    for key := range left_map {
        similarity += (key * left_map[key] * right_map[key])
    }

    fmt.Println("V2:", similarity)
}

File: day-1/rust/src/main.rs
use std::{collections::HashMap, fs};
#[derive(Debug)]
struct Input {
    left: Vec<String>,
    right: Vec<String>,
}

fn main() {
    v1();
    v2()
}

fn v1() {
    let text = fs::read_to_string("../input.txt").unwrap();
    let mut input = Input {
        left: Vec::new(),
        right: Vec::new(),
    };

    let _ = text.split("\n").for_each(|line| {
        let slices: Vec<&str> = line.split("   ").collect();
        if slices.len() == 2 {
            input.left.push(slices[0].to_string());
            input.right.push(slices[1].to_string());
        }
    });

    let distance = calc_distance(input);

    println!("V1: {}", distance)
}

fn calc_distance(mut input: Input) -> i32 {
    let mut distance = 0;

    input.left.sort();
    input.right.sort();

    for i in 0..input.left.len() {
        let left_number: i32 = input.left[i].parse().unwrap();
        let right_number: i32 = input.right[i].parse().unwrap();

        distance += (left_number - right_number).abs()
    }

    distance
}

fn v2() {
    let text = fs::read_to_string("../input.txt").unwrap();
    let mut left: HashMap<i32, i32> = HashMap::new();
    let mut right: HashMap<i32, i32> = HashMap::new();

    let _ = text.split("\n").for_each(|line| {
        let slices: Vec<i32> = line
            .split("   ")
            .map(|number| number.parse::<i32>().unwrap_or_default())
            .collect();
        if slices.len() == 2 {
            if let Some(value) = left.get_mut(&slices[0]) {
                *value += 1
            } else {
                left.insert(slices[0], 1);
            }
            if let Some(value) = right.get_mut(&slices[1]) {
                *value += 1
            } else {
                right.insert(slices[1], 1);
            }
        }
    });

    let mut similarity = 0;

    left.keys().for_each(|key| {
        let left_key = left.get(key).unwrap();
        let right_key = match right.get(key) {
            Some(value) => value,
            None => return (),
        };

        similarity += key * left_key * right_key;
    });

    println!("V2: {}", similarity)
}
